File: sample-chat-obj-c/sample-chat/Components/Authorization/RootParentVC.m
//
//  RootParentVC.m
//  samplechat
//
//  Created by Injoit on 1/29/20.
//  Copyright © 2020 Quickblox. All rights reserved.
//

#import "RootParentVC.h"
#import "SplashScreenVC.h"
#import "DialogsViewController.h"
#import "UIColor+Chat.h"

@interface DialogsNavigationController: UINavigationController
@end
@implementation DialogsNavigationController
@end

@interface RootParentVC ()
//MARK: - Properties
@property (nonatomic, strong) UIViewController *current;
@end

@implementation RootParentVC
//MARK: - Life Cycle
- (instancetype)init {
    self = [super init];
    if (self) {
        UIStoryboard *storyboard = [UIStoryboard storyboardWithName:@"Authorization" bundle:nil];
        SplashScreenVC *splashScreenVC = [storyboard instantiateViewControllerWithIdentifier:@"SplashScreenVC"];
        self.current = splashScreenVC;
    }
    return self;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    
    [self setupChildViewController:self.current];
}

//MARK - Setup
- (void)setDialogID:(NSString *)dialogID {
    _dialogID = dialogID;
    if (_dialogID) {
        [self handlePush];
    }
}

- (void)setupChildViewController:(UIViewController *)new {
    
    if ([self.current isEqual:new]) {
        [self addChildViewController:new];
        new.view.frame = self.view.bounds;
        [self.view addSubview:new.view];
        [new didMoveToParentViewController:self];

    } else {
        [self addChildViewController:new];
        new.view.frame = self.view.bounds;
        [self.view addSubview:new.view];
        [new didMoveToParentViewController:self];
        
        [self.current willMoveToParentViewController:nil];
        [self.current.view removeFromSuperview];
        [self.current removeFromParentViewController];
        self.current = new;
    }
}

- (UIStatusBarStyle) preferredStatusBarStyle {
    return UIStatusBarStyleLightContent;
}

// MARK: - Public Methods
- (void)showLoginScreen {
    UIStoryboard *storyboard = [UIStoryboard storyboardWithName:@"Authorization" bundle:nil];
    UINavigationController *new = [storyboard instantiateViewControllerWithIdentifier:@"AuthNavVC"];
    
    [self setupChildViewController:new];
}

- (void)showDialogsScreen {
    UIStoryboard *storyboard = [UIStoryboard storyboardWithName:@"Dialogs" bundle:nil];
    DialogsViewController *dialogsScreen = [storyboard instantiateViewControllerWithIdentifier:@"DialogsViewController"];
    
    DialogsNavigationController *new = [[DialogsNavigationController alloc] initWithRootViewController:dialogsScreen];
    new.navigationBar.barTintColor = [UIColor mainColor];
    new.navigationBar.barStyle = UIBarStyleBlack;
    new.navigationBar.shadowImage = UIImage.new;
    [new.navigationBar setTranslucent:NO];
    new.navigationBar.tintColor = UIColor.whiteColor;
    new.navigationBar.titleTextAttributes = @{ NSForegroundColorAttributeName:UIColor.whiteColor};
    
    [self setupChildViewController:new];
    
    [self handlePush];
}

//MARK: - Internal Methods
- (void)handlePush {
    if ([self.current isKindOfClass:[DialogsNavigationController class]] && self.dialogID) {
        DialogsNavigationController *dialogsNavigationController = (DialogsNavigationController *)self.current;
        [dialogsNavigationController popToRootViewControllerAnimated:NO];
        if ([dialogsNavigationController.topViewController isKindOfClass:[DialogsViewController class]]) {
            DialogsViewController *dialogsVC = (DialogsViewController *)dialogsNavigationController.topViewController;
            [dialogsVC openChatWithDialogID: self.dialogID];
            
            self.dialogID = nil;
        }
    }
}

@end

File: sample-chat-swift/sample-chat-swift/Components/Authorization/RootParentVC.swift
//
//  RootParentVC.swift
//  sample-chat-swift
//
//  Created by Injoit on 1/27/20.
//  Copyright © 2020 quickBlox. All rights reserved.
//

import UIKit

class DialogsNavigationController: UINavigationController { }

class RootParentVC: UIViewController {
    //MARK: - Properties
    var current: UIViewController
    var dialogID: String? {
        didSet {
            handlePush()
        }
    }
    
    //MARK: - Life Cycle
    init() {
        let storyboard = UIStoryboard(name: "Authorization", bundle: nil)
        current = storyboard.instantiateViewController(withIdentifier: "SplashScreenVC") as! SplashScreenVC
        super.init(nibName:  nil, bundle: nil)
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        changeCurrentViewController(current)
    }
    
    // MARK: - Public Methods
    func showLoginScreen() {
        let storyboard = UIStoryboard(name: "Authorization", bundle: nil)
        let authNavVC = storyboard.instantiateViewController(withIdentifier: "AuthNavVC") as! UINavigationController
        
        changeCurrentViewController(authNavVC)
    }
    
    func goToDialogsScreen() {
        let storyboard = UIStoryboard(name: "Dialogs", bundle: nil)
        if let dialogsVC = storyboard.instantiateViewController(withIdentifier: "DialogsViewController") as? DialogsViewController {
            let dialogsScreen = DialogsNavigationController(rootViewController: dialogsVC)
            dialogsScreen.navigationBar.barTintColor = #colorLiteral(red: 0.2216441333, green: 0.4713830948, blue: 0.9869660735, alpha: 1)
            dialogsScreen.navigationBar.barStyle = .black
            dialogsScreen.navigationBar.shadowImage = UIImage()
            dialogsScreen.navigationBar.isTranslucent = false
            dialogsScreen.navigationBar.tintColor = .white
            dialogsScreen.navigationBar.titleTextAttributes = [NSAttributedString.Key.foregroundColor: UIColor.white]
            
            changeCurrentViewController(dialogsScreen)

            handlePush()
        }
    }
    
    //MARK: - Internal Methods
    private func changeCurrentViewController(_ newCurrentViewController: UIViewController) {
            addChild(newCurrentViewController)
            newCurrentViewController.view.frame = view.bounds
            view.addSubview(newCurrentViewController.view)
            newCurrentViewController.didMove(toParent: self)
            
            if current == newCurrentViewController {
                return
            }
            current.willMove(toParent: nil)
            current.view.removeFromSuperview()
            current.removeFromParent()
            current = newCurrentViewController
        }
    
    private func handlePush() {
        if let dialogsNavigationController = current as? DialogsNavigationController, let dialogID = dialogID {
            if let dialog = ChatManager.instance.storage.dialog(withID: dialogID) {
                // Autojoin to the group chat
                if dialog.type != .private, dialog.isJoined() == false {
                    dialog.join(completionBlock: { error in
                        guard let error = error else {
                            return
                        }
                        debugPrint("[RootParentVC] dialog.join error: \(error.localizedDescription)")
                    })
                }
                dialogsNavigationController.popToRootViewController(animated: false)
                (dialogsNavigationController.topViewController as? DialogsViewController)?.openChatWithDialogID(dialogID)
                self.dialogID = nil
            }
        }
    }
    
    //MARK - Setup
    override var preferredStatusBarStyle: UIStatusBarStyle {
        return .lightContent
    }
}
